### PR TITLE
Initial fake juju CLI skeleton

### PR DESCRIPTION
**Pull request by freeekanayaka**
Add a new fake-juju command and associated Go source and make
target. This command will behave exactly like the regular 'juju' cli
tool, except that it will always point to a fake-jujud instance.

This branch adds the initial code to write files to the JUJU_DATA
directory, pointing to a fake-jujud controller.

Integration tests have been added to exercise this initial behavior,
and assert that the expected files are actually written in the JUJU_DATA
directory.

Follow-up branches will further change the bootstrap command to
connect to fake-juju and perform a fake "bootstrap" by creating
and starting the controller machine (aka machine 0).

### CODE REVIEW

File: common/core/github.com/juju/juju/cmd/juju/commands/bootstrap-fakejuju.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/juju/juju/version"
+)
+
+// Custom bootstrap logic for fake juju, which essentially:
+//
+// - populates the JUJU_DATA directory with files pointing to the
+//   fake-jujud process, as identified by the port number in the
+//   FAKE_JUJUD_PORT environment variable.
+//
+// - connects to fake-jujud an get the agent version to ensure that
```
Comment (dsegan) [inline common/core/github.com/juju/juju/cmd/juju/commands/bootstrap-fakejuju.go]: "connects to fake-jujud *and gets* the agent..."

File: common/core/github.com/juju/juju/cmd/juju/commands/bootstrap-fakejuju.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	if err := writeAccountsFile(store, controller); err != nil {
+		return err
+	}
+
+	if err := writeModelsFile(store, controller, model); err != nil {
+		return err
+	}
+
+	if err := store.SetCurrentModel(controller, model); err != nil {
+	}
```
Comment (dsegan) [inline common/core/github.com/juju/juju/cmd/juju/commands/bootstrap-fakejuju.go]: Did you want to do the same as all the above if blocks? If not, perhaps not even check for err and just call store.SetCurrentModel() directly.

File: common/makefile.mk [hunk 3]
```diff
@@ -22,10 +22,19 @@ JUJU_TARBALL_URL=https://launchpad.net/$(JUJU_PROJECT)/$(JUJU_MAJOR_MINOR)/$(JUJ
 
 .PHONY: build
 build: $(JUJU_TARBALL) $(JUJU_PATCH)
+
 	rm -rf $(JUJU_SRC)  # Go doesn't play nice with existing files.
+
+	# Extract the original tarball, apply the fake-juju patch and create
+	# synlinks in the original tree for the additional fake-juju sources.
```
Comment (dsegan) [inline common/makefile.mk]: "syMlinks" and I believe "the" in "the additional fake-juju sources" is not needed

**Comment by freeekanayaka** [comment]
Danilo Šegan <notifications@github.com> writes:

> dsegan approved this pull request.
>
> Minor comments inline, looks good.
>
>> +	"strconv"
> +
> +	"github.com/juju/errors"
> +
> +	"github.com/juju/juju/jujuclient"
> +	"github.com/juju/juju/testing"
> +	"github.com/juju/juju/version"
> +)
> +
> +// Custom bootstrap logic for fake juju, which essentially:
> +//
> +// - populates the JUJU_DATA directory with files pointing to the
> +//   fake-jujud process, as identified by the port number in the
> +//   FAKE_JUJUD_PORT environment variable.
> +//
> +// - connects to fake-jujud an get the agent version to ensure that
>
> "connects to fake-jujud *and gets* the agent..."

Fixed.

>
>> +	model := "controller"
> +
> +	if err := writeControllersFile(store, controller); err != nil {
> +		return err
> +	}
> +
> +	if err := writeAccountsFile(store, controller); err != nil {
> +		return err
> +	}
> +
> +	if err := writeModelsFile(store, controller, model); err != nil {
> +		return err
> +	}
> +
> +	if err := store.SetCurrentModel(controller, model); err != nil {
> +	}
>
> Did you want to do the same as all the above if blocks? If not,
> perhaps not even check for err and just call store.SetCurrentModel()
> directly.

Oops. Yes, I wanted to propagate the error as all above. Good catch :)

**Pull request merged**